File: 백준/Gold/19598. 최소 회의실 개수/최소 회의실 개수.java
import java.util.*;
import java.io.*;

public class Main {

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int N = Integer.parseInt(br.readLine());
        int[][] time = new int[N][2]; // 강의 시작과 끝 시간 저장 배열
        PriorityQueue<Integer> room = new PriorityQueue<>();

        for(int i = 0; i < N; i++) {
            time[i] = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        }
        // 시작 순서 오름차순, 종료 순서 오름차순 정렬
        Arrays.sort(time, (o1, o2) -> {
            if(o1[0] == o2[0]) {
                return o1[1] - o2[1];
            }
            else {
                return o1[0] - o2[0];
            }
        });

        br.close();
        room.add(time[0][1]);
        // 시간대가 겹치지 않으면 room.poll(), 시간대가 겹치면 room.add()만 해서 강의실 추가
        for(int i = 1; i < N; i++) {
            if(room.peek() <= time[i][0]) {
                room.poll();
            }
            room.add(time[i][1]);
        }
        bw.write(room.size()+"");
        bw.flush();
        bw.close();
    }
}

File: 백준/Gold/19598. 최소 회의실 개수/최소 회의실 개수.kt
import java.util.*
import java.io.*

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {

    val bw = BufferedWriter(OutputStreamWriter(System.out))
    val N = readLine().toInt()
    val meeting = Array(N){IntArray(2)} // 회의 정보 저장
    val pq = PriorityQueue<Int>()

    for(i in 0 until N) {
        val (start, end) = readLine().split(" ").map { it.toInt() }
        meeting[i][0] = start
        meeting[i][1] = end
    }
    close()

    // 회의 시작 시간 오름차순, 종료 시간 오름차순 정렬
    meeting.sortWith(compareBy({it[0]}, {it[1]}))
    pq.offer(meeting[0][1])

    // pq에는 회의 종료 시간을 저장
    for(i in 1 until N) {
        // 이전 회의 종료와 다음 회의 시작이 겹치지 않으면 pq.poll()
        if(pq.peek() <= meeting[i][0]) {
            pq.poll()
        }
        pq.offer(meeting[i][1])
    }
    bw.write(pq.size.toString())
    bw.flush()
    bw.close()
}